cmd/test/main.go

```diff
@@ -0,0 +1,368 @@
+	dbName := "tst_" + ravendb.NewUUID().Hex()
+	exec := getExecutor()
+	cmd := ravendb.NewCreateDatabaseCommand(dbName, 1)
+	res, err := ravendb.ExecuteCreateDatabaseCommand(exec, cmd)
+	must(err)
+	panicIf(res.RaftCommandIndex == 0, "res.RaftCommandIndex is 0")
+	panicIf(res.Name != dbName, "res.Name is '%s', expected '%s'", res.Name, dbName)
+	if true || verboseLog {
+		fmt.Printf("res: %#v\n", res)
+	}
+
+	// TODO: do I need to wait?
+
+	cmd2 := ravendb.NewDeleteDatabaseCommand(dbName, false, "")
+	res2, err := ravendb.ExecuteDeleteDatabaseCommand(exec, cmd2)
+	must(err)
+	panicIf(res2.RaftCommandIndex == 0, "res2.RaftCommandIndex is 0")
+	if verboseLog {
+		fmt.Printf("res2: %#v\n", res2)
+	}
+
+	fmt.Printf("testCreateAndDeleteDatabaseCommand ok\n")
+}
+*/
+
+func testPutGetDeleteDocument() {
+	sess := mustGetSession()
+	exec := sess.RequestExecutor.GetCommandExecutor(false)
+	key := "testing/1"
+	meta := map[string]interface{}{
+		"@collection": "Testings",
+	}
+	doc := map[string]interface{}{
+		"Name":      "test1",
+		"DocNumber": 1,
+		"@metadata": meta,
+	}
+	cmd := ravendb.NewPutDocumentRawCommand(key, doc, "")
+	res, err := ravendb.ExecutePutDocumentRawCommand(exec, cmd)
+	must(err)
+	if verboseLog {
+		fmt.Printf("res: %#v\n", res)
+	}
+
+	cmd = ravendb.NewGetDocumentCommand([]string{key}, nil, false)
+	res2, err := ravendb.ExecuteGetDocumentCommand(exec, cmd)
+	must(err)
+	if verboseLog {
+		fmt.Printf("len(res2.Includes): %d, len(res2.Results): %d\n", len(res2.Includes), len(res2.Results))
+	}
+
+	// test get of non-existent document
+	cmd = ravendb.NewGetDocumentCommand([]string{"testing/asdf"}, nil, false)
+	res3, err := ravendb.ExecuteGetDocumentCommand(exec, cmd)
+	panicIf(err == nil, "unexpected err is nil")
+	panicIf(res3 != nil, "unexpected res3 is != nil: %#v", res3)
+	// verify that it returns 404 Not Found
+	notFound := err.(*ravendb.NotFoundError)
+	if verboseLog {
+		fmt.Printf("not found url: '%s'\n", notFound.URL)
+	}
+
+	cmd = ravendb.NewDeleteDocumentCommand(key, "")
+	err = ravendb.ExecuteDeleteDocumentCommand(exec, cmd)
+	must(err)
+
+	// test delete of non-existent document
+	// it succeeds even if document doesn't exist
+	cmd = ravendb.NewDeleteDocumentCommand("testing/asdf", "")
+	err = ravendb.ExecuteDeleteDocumentCommand(exec, cmd)
+	must(err)
+
+	// TODO: test changeVector
+	fmt.Printf("testPutGetDeleteDocument ok\n")
+}
+
+func testHiLoKeyGenerator() {
+	store := getStore()
+	tag := "my_tag"
+	generator := ravendb.NewHiLoKeyGenerator(tag, store, testDbName)
+	fmt.Printf("generator: %#v\n", generator)
+	res := generator.GenerateDocumentKey()
+	if verboseLog {
+		fmt.Printf("%#v\n", res)
+	}
+	err := generator.ReturnUnusedRange()
+	must(err)
+	print("testHiLoKeyGenerator ok\n")
+}
+
+// Foo is just a test object
+type Foo struct {
+	ID   string
+	Name string
+}
+
+func testStoreLoad() {
+	sess := mustGetSession()
+	v := &Foo{
+		Name: "PyRavenDB",
+	}
+	err := sess.Store(v, "", "")
```

Comment:
I know that `Go` doesn't support overloading / optional parameters.
However, this gives rise to bad semantics. `""` as the change vector means "enforce new", but the usual way we do it elsewhere is with null and overwrite by default.
I don't think users will be very happy about this subtle distinction, and it is likely to trip them up.

What would be the `Go` way to resolve this? Different method names, I assume? `Store(doc)` `StoreId(doc, id)`, etc?

Reply:
There are two options.

Like you said, additional functions. Go naming would use "With". I already have some of those in InMemoryDocumentSessionOperations` e.g. `Delete` and `DeleteWithChangeVector`.

If there are too many arguments, we can create an Arg struct combining multiple optional args and pass that in.